src/server.cpp

```diff
@@ -1575,12 +1575,13 @@ void CServer::FreeChannel ( const int iCurChanID )
             --iCurNumChannels;
 
             // move channel IDs down by one starting at the bottom and working up
-            while ( i < iCurNumChannels )
+            // also keep the free channels in sorted order by moving lower-numbered ones down
+            while ( i < iCurNumChannels || ( i + 1 < iMaxNumChannels && vecChannelOrder[i + 1] < iCurChanID ) )
             {
                 int j              = i++;
                 vecChannelOrder[j] = vecChannelOrder[i];
             }
-            // put deleted channel at the end ready for re-use
+            // put deleted channel in the correct position ready for re-use
```

Comment:
This all now seems very cumbersome.  If it's going to sit _somewhere_ in `vecChannelOrder`, why not just leave it where it was?

Reply:
Because the binary search relies on all the active channels being listed before all the free ones. All we are changing is how many IDs we shunt down the list when freeing, so that the freed ID is placed in order instead of just at the start of the free list. Doesn't seem cumbersome to me at all :)

Reply:
So you're saying it was (allocated | free):
```
1 2 3 4 5   ||
1 3 4 5 | 2 ||
1 3 4 | 5 2 ||
3 4 | 1 5 2 ||
```
and now it's
```
1 2 3 4 5   ||
1 3 4 5 | 2 ||
1 3 4 | 2 5 ||
3 4 | 1 2 5 ||
```
?  I think the comment doesn't explain anything.

Reply:
Yes, exactly that. I'll change the comment to say "put deleted channel in the _vacated_ position ready for re-use"